### PR TITLE
Refactor sl input enabled mixin

### PR DESCRIPTION
**Pull request by azizpunjani**

### CODE REVIEW

**Comment by azizpunjani** [comment]
Fixes#https://github.com/softlayer/sl-ember-components/issues/791

File: addon/templates/components/sl-input.hbs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 {{/if}}
 
 {{input
+    id=inputId
```
Comment (notmessenger) [inline addon/templates/components/sl-input.hbs]: Why is this added when it is removed from sl-radio and sl-select?

File: addon/mixins/sl-component-input-id.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    // -------------------------------------------------------------------------
+    // Actions
+
+    // -------------------------------------------------------------------------
+    // Events
+
+    // -------------------------------------------------------------------------
+    // Properties
+
```
Comment (notmessenger) [inline addon/mixins/sl-component-input-id.js]: Define the `inputId` property

File: tests/integration/components/sl-input-test.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForComponent( 'sl-input', 'Integration | Component | sl input', {
+    integration: true
+});
+
+test( 'for attribute value on label matches id of input', function( assert ) {
+    this.render( hbs`
+        {{sl-input label='test label'}}
```
Comment (notmessenger) [inline tests/integration/components/sl-input-test.js]: Use double quotes

File: tests/integration/components/sl-select-test.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        {{sl-select label='test label'}}
+    ` );
+
+    const wrapper = this.$( '>:first-child' );
+
+    assert.equal(
+        wrapper.find( 'label' ).attr( 'for' ),
+        wrapper.find( 'input' ).attr( 'id' )
+    );
+});
```
Comment (notmessenger) [inline tests/integration/components/sl-select-test.js]: I thought we were not going to create this file as we have no other tests for this component and have no plans for any until https://github.com/softlayer/sl-ember-components/issues/456

File: tests/integration/components/sl-select-test.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForComponent( 'sl-select', 'Integration | Component | sl select', {
+    integration: true
+});
+
+test( 'for attribute value on label matches id of input', function( assert ) {
+    this.render( hbs`
+        {{sl-select label='test label'}}
```
Comment (notmessenger) [inline tests/integration/components/sl-select-test.js]: Use double quotes

**Comment by notmessenger** [comment]
The unit tests of the affected components need an assertion that the `sl-component-input-id` mixin is being mixed in.

File: tests/integration/components/sl-textarea-test.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForComponent( 'sl-textarea', 'Integration | Component | sl textarea', {
+    integration: true
+});
+
+test( 'for attribute value on label matches id of textarea', function( assert ) {
+    this.render( hbs`
+        {{sl-textarea label='test label'}}
```
Comment (notmessenger) [inline tests/integration/components/sl-textarea-test.js]: Use double quotes

**Comment by notmessenger** [comment]
Do the `sl-radio`, `sl-date-picker`, and `sl-date-range-picker` also need this same treatment?

File: addon/templates/components/sl-input.hbs (hunk 1 above)
Comment (azizpunjani) [inline addon/templates/components/sl-input.hbs]: The input within the `sl-radio` component doesn't need an id assigned because the label wraps the input, hence a for attribute it not required on the label.

`sl-select` component hides the original input provided in the template and creates it's own input within a div. I was able to assign the id after the element renders in this code https://github.com/softlayer/sl-ember-components/pull/798/files#diff-6e21ecc6b00cadbe559301f6b81a8b3eR280.

File: tests/integration/components/sl-select-test.js (hunk 4 above)
Comment (azizpunjani) [inline tests/integration/components/sl-select-test.js]: Since I was able to assign the input an id generated in the mixin, I figured I would add the test in for it, the test passes fine.

**Comment by azizpunjani** [comment]
@notmessenger yes `sl-date-picker` and `sl-date-range-picker` need the same code changes. They will be handled by @juwara0 in issues https://github.com/softlayer/sl-ember-components/issues/550 and https://github.com/softlayer/sl-ember-components/issues/553.

**Comment by azizpunjani** [comment]
@notmessenger this is ready for review. I have fixed issues you mentioned.

File: addon/templates/components/sl-select.hbs [hunk 7]
```diff
@@ -1,5 +1,5 @@
 {{#if label}}
-    <label for={{inputId}} class="control-label">
```
Comment (azizpunjani) [inline addon/templates/components/sl-select.hbs]: Removed this as it's currently inaccurate.

File: addon/templates/components/sl-select.hbs [hunk 8]
```diff
@@ -13,7 +13,6 @@
 {{/if}}
 
 {{input
-    id=inputId
```
Comment (azizpunjani) [inline addon/templates/components/sl-select.hbs]: Removed this as it's currently inaccurate.

File: addon/mixins/sl-component-input-id.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Unique input id that a component can assign to an input
+     * and a labels for attribute
+     *
+     * @type {String}
```
Comment (notmessenger) [inline addon/mixins/sl-component-input-id.js]: Missing `?'

File: addon/mixins/sl-component-input-id.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    // -------------------------------------------------------------------------
+    // Events
+
+    // -------------------------------------------------------------------------
+    // Properties
+
+
+    /**
+     * Unique input id that a component can assign to an input
+     * and a labels for attribute
```
Comment (notmessenger) [inline addon/mixins/sl-component-input-id.js]: Add possessive apostrophe to "labels"

File: addon/templates/components/sl-input.hbs (hunk 1 above)
Comment (notmessenger) [inline addon/templates/components/sl-input.hbs]: RESOLVED

File: addon/mixins/sl-component-input-id.js (hunk 2 above)
Comment (notmessenger) [inline addon/mixins/sl-component-input-id.js]: RESOLVED

**Comment by notmessenger** [comment]
Does this https://github.com/softlayer/sl-ember-components/pull/798/files#diff-2d71ed7147645907b16a78a7ebcf95faR12 have an advantage over https://github.com/softlayer/sl-ember-components/blob/master/tests/unit/mixins/sl-input-based-test.js#L11 and https://github.com/softlayer/sl-ember-components/blob/master/tests/unit/mixins/sl-tooltip-enabled-test.js#L37 ?  If not, make it consistent with these two.  If so, create issues to change these occurrences so codebase remains consistent in its usage of patterns

File: addon/mixins/sl-component-input-id.js (hunk 9 above)
Comment (azizpunjani) [inline addon/mixins/sl-component-input-id.js]: Yes, fixed.

**Comment by notmessenger** [comment]
Do the component integration tests need to have test for dual rendering in template, to ensure that ids are not be cross-populated?

**Comment by azizpunjani** [comment]
@notmessenger 

>Does this https://github.com/softlayer/sl-ember-components/pull/798/files#diff- 2d71ed7147645907b16a78a7ebcf95faR12 have an advantage over https://github.com/softlayer/sl-ember-components/blob/master/tests/unit/mixins/sl-input-based-test.js#L11 and https://github.com/softlayer/sl-ember-components/blob/master/tests/unit/mixins/sl-tooltip-enabled-test.js#L37 ? If not, make it consistent with these two. If so, create issues to change these occurrences so codebase remains consistent in its usage of patterns.

Yes, using mixin.detect is a more accurate test because it detects the presence of the mixin within the Object/Class. I have opened an issue to update existing code https://github.com/softlayer/sl-ember-components/issues/804

**Comment by azizpunjani** [comment]
@notmessenger 

>Does this https://github.com/softlayer/sl-ember-components/pull/798/files#diff-2d71ed7147645907b16a78a7ebcf95faR12 have an advantage over https://github.com/softlayer/sl-ember-components/blob/master/tests/unit/mixins/sl-input-based-test.js#L11 and https://github.com/softlayer/sl-ember-components/blob/master/tests/unit/mixins/sl-tooltip-enabled-test.js#L37 ? If not, make it consistent with these two. If so, create issues to change these occurrences so codebase remains consistent in its usage of patterns.

#737 also needs to be in sync.

I think it makes sense to use `mixin.detect` within component unit tests to see if a mixin is present within a component instance. Within the mixin unit test, to check if a mixin was mixed in successfully, it makes sense to check if the object instance was created successfully after extending the mixin.

**Comment by notmessenger** [comment]
Shouldn't https://github.com/azizpunjani/sl-ember-components/commit/3ffc0242e6acaadb978e4816f6b4ec22bb3642ae be reverted, to align with the reversion in https://github.com/azizpunjani/sl-ember-components/commit/0e908fb19fc892bcde4d7e98a942781f390cef50 ?
